File: typescript/day2/src/day2.ts
import { interpretLineAsActionAndOutcome, interpretLineAsTwoActions } from './core_function.ts'
import { exists } from 'fs/exists.ts'

const args = Deno.args
if (args.length !== 1) {
  console.log(`Expected 3 arguments, received ${args.length}`)
  Deno.exit(1)
}

if (!exists(args[0])) {
  console.log(`The file ${args[0]} does not exist`)
  Deno.exit(1)
}

const start = performance.now()
const decoder = new TextDecoder('utf-8')
const input = decoder.decode(Deno.readFileSync(args[0]))

const scoreIfLinesAreActions = input.split('\n')
  .filter((str) => str.length > 0)
  .map(interpretLineAsTwoActions)
  .reduce((sum, current) => {
    return sum + current
  }, 0)

const scoreIfLinesAreActionAndOutcome = input.split('\n')
  .filter((str) => str.length > 0)
  .map(interpretLineAsActionAndOutcome)
  .reduce((sum, current) => {
    return sum + current
  }, 0)

console.log(`Score if each line is two actions: ${scoreIfLinesAreActions}`)
console.log(`Score if each line is an action and outcome: ${scoreIfLinesAreActionAndOutcome}`)
console.log(`Elapsed time: ${(performance.now() - start).toPrecision(5)} milliseconds`)
Deno.exit(0)

File: go/day2/day2.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"time"
)

func interpretLineAsTwoActions(line string) (int, error) {
	switch line {
	case "A X":
		return 4, nil
	case "A Y":
		return 8, nil
	case "A Z":
		return 3, nil
	case "B X":
		return 1, nil
	case "B Y":
		return 5, nil
	case "B Z":
		return 9, nil
	case "C X":
		return 7, nil
	case "C Y":
		return 2, nil
	case "C Z":
		return 6, nil
	default:
		return -1, errors.New("unrecognized line")
	}
}

func interpretLineAsActionAndOutcome(line string) (int, error) {
	switch line {
	case "A X":
		return 3, nil
	case "A Y":
		return 4, nil
	case "A Z":
		return 8, nil
	case "B X":
		return 1, nil
	case "B Y":
		return 5, nil
	case "B Z":
		return 9, nil
	case "C X":
		return 2, nil
	case "C Y":
		return 6, nil
	case "C Z":
		return 7, nil
	default:
		return -1, errors.New("unrecognized line")
	}
}

func scoreInput(scanner *bufio.Scanner) (int, int) {
	score_if_actions, score_if_action_and_outcome := 0, 0
	for scanner.Scan() {
		score, err := interpretLineAsTwoActions(scanner.Text())

		if err == nil {
			score_if_actions += score
		}

		score, err = interpretLineAsActionAndOutcome(scanner.Text())
		if err == nil {
			score_if_action_and_outcome += score
		}
	}

	return score_if_actions, score_if_action_and_outcome
}

func main() {
	if len(os.Args) != 2 {
		fmt.Println("Wrong number of arguments, expected filename.")
		os.Exit(1)
	}

	input_file, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	defer input_file.Close()

	start := time.Now()
	scanner := bufio.NewScanner(input_file)
	score_if_actions, score_if_action_and_outcome := scoreInput(scanner)

	fmt.Println("my score if the lines are actions: ", score_if_actions)
	fmt.Println("my score if the lines are action and outcome: ", score_if_action_and_outcome)
	fmt.Printf("Calculated in: %v\n", time.Since(start))
}
